### ISSUE TITLE
Header generation not conforming to RFC822 and RFC2046

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2346]
|[<img alt="teknico's avatar" src="https://avatars.githubusercontent.com/u/172521?s=50" width="50" height="50">](https://github.com/teknico)| @teknico reported|
|-|-|
|Trac ID|trac#2346|
|Type|defect|
|Created|2007-01-02 16:47:18Z|
|Branch|https://github.com/twisted/twisted/tree/header-value-quoting-2346|

Currently, header generation in web2 does not quote parameter values, as in ```"Content-Type: text/plain; name=foo"``` .
This breaks when the value contains token specials (eg. "="), in which case RFC2046 recommends quoting anyway.

The attached test patch (```test_http_headers.diff```) causes a test failure in web2.http_headers , that is fixed by the other patch (```http_headers.diff```).

Attachments:

* [test_http_headers.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a6e/a6e30b2e057ef7205fde876d0be2f643aa74c0cb/7fb631ce4ddc045afd6c50844116ee3749fcdf2c.diff) (855 bytes) - added by tekNico on 2007-01-02 16:47:42Z - Test patch demonstrating the problem
* [http_headers.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a6e/a6e30b2e057ef7205fde876d0be2f643aa74c0cb/90ff4d01f7363ec1ad768005346e425b1c9c2342.diff) (1778 bytes) - added by tekNico on 2007-01-02 16:52:54Z - Patch fixing the problem
* [http_headers_2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a6e/a6e30b2e057ef7205fde876d0be2f643aa74c0cb/2264d0bafe1db9092f731209d958ecff9decc203.diff) (1967 bytes) - added by tekNico on 2007-01-03 11:41:23Z - Reworked fix patch
* [test_http_headers_2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a6e/a6e30b2e057ef7205fde876d0be2f643aa74c0cb/4e1324f3da34ad0bc9dac39cf1d60addc5ad03f4.diff) (1114 bytes) - added by tekNico on 2007-01-03 11:41:44Z - Reworked test patch

<details><summary>Searchable metadata</summary>

```
trac-id__2346 2346
type__defect defect
reporter__tekNico tekNico
priority__normal normal
milestone__ 
branch__branches_header_value_quoting_2346 branches/header-value-quoting-2346
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1167756438000000 1167756438000000
changetime__1225232740000000 1225232740000000
version__None None
owner__ 
cc__tekNico cc__dvd@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_1"></a>|@jyknight commented|
|-|-|

I have this recollection that there are prominent apps/servers out there which don't handle quotes in the boundary string properly (that is, they will look for a literal ```--"===============1318316892"``` line.

So, it must not output quotes unless actually necessary, and really, nobody should use boundary strings which require quotes if they want to interoperate.

**Comment by twisted-trac** [comment]
|[<img alt="teknico's avatar" src="https://avatars.githubusercontent.com/u/172521?s=50" width="50" height="50">](https://github.com/teknico)<a name="note_2"></a>|@teknico commented|
|-|-|

Yes, always quoting, even if allowed by the RFC, may be problematic. On the other hand, it's important to quote when the boundary contains special characters (that's what you meant by "actually necessary", right?), and web2 currently does not do that.

We reworked the fix patch with that approach, using the ```quoteString``` function and the ```tspecials``` regular expression from the email standard package. By the way, the email package is the one originating this problem, because it puts equal-sign characters in the boundary parameters it generates. 

We also reworked the test patch, adding a check for a boundary without quotes.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [25169]) Branching to 'header-value-quoting-2346'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [25170]) do quoting

refs [#2346](https://github.com/twisted/twisted/issues/2346)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Implementing something like what's in the latest patches, but with more fine-grained tests and a bunch of references to the relevant RFCs, plus a fix (the RFC mandates quoting HT, {, and } which the patch wasn't quoting).

Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [25171]) use http_tokens, a pre-existing definition of the rfc 2616 separators production

refs [#2346](https://github.com/twisted/twisted/issues/2346)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [25172]) no need to make it into a string again (although I think str.join special-cases this)

refs [#2346](https://github.com/twisted/twisted/issues/2346)

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_10"></a>|@jyknight set owner to @exarkun|
|-|-|

+1 looks good to me

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [25173]) Merge header-value-quoting-2346

Author: exarkun, itamar
Reviewer: jknight
Fixes: [#2346](https://github.com/twisted/twisted/issues/2346)

Fix generation of HTTP header values in web2 for the case where a parameter value
includes a byte which requires quoting.  Previously these values would not be
quoted, resulting in output which probably could not be parsed.  Now quoting will
be applied if necessary to avoid generating bad output.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
